### PR TITLE
Alpha-Candidates: Improve the performance when filtering search.

### PR DESCRIPTION
**Pull request by alwayslove2013**
issue: #5 

Adjusted the hnsw algorithm so that invalid nodes will enter the `candidates` with an alpha probability rather than 100%.

As shown in the figure, at high filtering rates, QPS performs better, and recall is also around 90%.

image
image

### CODE REVIEW

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 1]
```diff
@@ -364,6 +366,12 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t> {
                 int status = Neighbor::kValid;
                 if (has_deletions && bitset.test((int64_t)v)) {
                     status = Neighbor::kInvalid;
+
+                    accumulative_alpha += alpha;
```
[the remark this answers is not in the document]
Comment (alwayslove2013) [reply thirdparty/hnswlib/hnswlib/hnswalg.h:364]: nice~

File: thirdparty/hnswlib/hnswlib/hnswalg.h [hunk 2]
```diff
@@ -41,6 +41,7 @@ typedef unsigned int tableint;
 typedef unsigned int linklistsizeint;
 constexpr float kHnswSearchKnnBFThreshold = 0.93f;
 constexpr float kHnswSearchRangeBFThreshold = 0.97f;
+constexpr float alpha = 0.15f;
```
Comment (PwzXxm) [inline thirdparty/hnswlib/hnswlib/hnswalg.h:44]: Maybe use `kAlpha` to following the code-style?

**Comment by chasingegg** [comment]
> issue: #5
> 
> Adjusted the hnsw algorithm so that invalid nodes will enter the `candidates` with an alpha probability rather than 100%.
> 
> As shown in the figure, at high filtering rates, QPS performs better, and recall is also around 90%.
> 
> image image

why the recall does not go down in the `knowhere base` in a large (around 90%) filter rate, it is surprisingly stable

**Comment by PwzXxm** [comment]
Need to update the performance chart.

**Pull request merged**